[ML] Return statistics about forecasts as part of the jobsstats and usage API

hendrikmuhs opened the pull request:
This change adds stats about forecasts, to the jobstats api as well as xpack/_usage. The following information is collected:

#### _xpack/ml/anomaly_detectors/{jobid|_all}/_stats:
 - total number of forecasts
 - memory statistics (mean/min/max)
 - runtime statistics
 - record statistics
 - counts by status

#### _xpack/usage
- collected by job status as well as overall (`_all`):
   - total number of forecasts
   - number of jobs that have at least 1 forecast 
   - memory, runtime, record statistics
   - counts by status

Fixes #31395

Implementation notes:

- reused existing functionality but had to move classes to core due to dependencies
- added functionality to use aggregations for the collection of statistics

Todo: 

- [x] check documentation/specs (https://www.elastic.co/guide/en/elasticsearch/reference/current/index.html)
- [ ] schema updates for consumers of above API's

elasticmachine commented:
Pinging @elastic/ml-core

hendrikmuhs commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/stats/StatsAccumulator.java:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Helper class to collect min, max, avg and total statistics for a quantity
+ */
+public class StatsAccumulator implements Writeable {
```
this has been moved from `o.e.x.ml.utils`

hendrikmuhs commented:
Examples:

#### _xpack/ml/anomaly_detectors/{job|_all}/_stats 

![screenshot_20180628_140143](https://user-images.githubusercontent.com/7126422/42033025-ca797862-7adb-11e8-9fc2-648ae1da2760.png)

#### _xpack/ml/anomaly_detectors/{job|_all}/_stats  no forecast

![screenshot_20180628_140354](https://user-images.githubusercontent.com/7126422/42033119-188b14c0-7adc-11e8-8e56-54f55b3ca1d7.png)

**Note**: No statistics, only the total counter showing `0` forecasts for this jobs

#### _xpack/usage

![screenshot_20180628_140533](https://user-images.githubusercontent.com/7126422/42033188-53ca091a-7adc-11e8-8716-87df2a1b8ca1.png)

**Note**: `jobs` (better name anyone?) is the unique count of jobs that have at least 1 forecast. E.g. in this example `count == 4`, `jobs == 3`, so 3 out of 4 ML jobs have at least 1 forecast
`jobs` is omitted if it is 1.

davidkyle commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/stats/ForecastStats.java:
```diff
@@ -0,0 +1,153 @@
+        }
+        total += other.total;
+        if (other.total > 0) {
+            ++jobsWithAtleastOneForecast;
```
`fs = new ForecastStats().merge(a.merge(b.merge(c.merge(someStats))))`
After this `fs. jobsWithAtleastOneForecast` would have a value of 1 regardless of what has been merged into should the value of `other. jobsWithAtleastOneForecast` be used or have I misunderstood this completely.

hendrikmuhs replied:
I think you are right, it's not used like this right now, but that's a bad excuse for not fixing it. it should simply be `jobsWithAtleastOneForecast +=other.jobsWithAtleastOneForecast`, the if is superflous.

davidkyle commented:
> Note: jobs (better name anyone?) is the unique count of jobs that have at least 1 forecast. E.g. in this example count == 4, jobs == 3, so 3 out of 4 ML jobs have at least 1 forecast
jobs is omitted if it is 1.

I didn't see this before I started the review. How about `forecasted_jobs`?

hendrikmuhs commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/stats/ForecastStats.java:
```diff
@@ -0,0 +1,153 @@
+            builder.field(Fields.RUNTIME, runtimeStats.asMap());
+            builder.field(Fields.STATUSES, statusCounts.asMap());
+        }
+        if (jobsWithAtleastOneForecast > 1) {
```
I agree, this magic is a bit strange. The reason is that I wanted to not show this field for the jobs stats API but only for the usage stats.

hendrikmuhs commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearningFeatureSet.java:
```diff
@@ -215,24 +219,29 @@ private void addJobsUsage(GetJobsStatsAction.Response response) {
                 modelSizeStatsByState.computeIfAbsent(jobState,
                         js -> new StatsAccumulator()).add(modelSize);
+                forecastStatsByState.computeIfAbsent(jobState,
+                        js -> new ForecastStats()).merge(jobStats.getForecastStats());
```
not sure if I get the question, only in case the key does not exist yet it is equivalent to `jobStats.getForecastStats()`. I think I can simplify this using the `Map::merge` function in this case.

hendrikmuhs commented:
I changed the fieldname as proposed to `forecasted_jobs` and removed the magic about writing it or not. It's now always written, for stats I still avoid writing them if there isn't any forecast.

hendrikmuhs commented:
@lcawl 

Would you mind quickly reviewing: https://github.com/elastic/elasticsearch/pull/31647/commits/f9804dca36d28db9dfe7bc3437675258e1724fda ?
